### DISCUSSION TITLE
Magma Forge + Plugin Support [FEATURE REQUEST]

### DISCUSSION THREAD

**Discussion by Tim-Rambo** [Zrips/CMI#3847] [Feature Requests]
**Description of feature/improvement request:**

---
### Detailed explanation of what it should do, when and in what situation: ###
Add support for magma so i can use the plugin with my favorite mods
### Basic explanation as to why this feature is needed and/or useful to have: ###
So we can use CMI with mods.

https://magmafoundation.org

**Comment by CloudeLecaw** [comment]
no ^,^

Ask the Magma Dev to fix his server fork to work with Spigot or PaperMC :P

**Comment by Tim-Rambo** [comment]
> no ^,^
> 
> Ask the Magma Dev to fix his server fork to work with Spigot or PaperMC :P

take this atleast.
https://github.com/magmafoundation/Magma/issues/204#issue-667786826

**Comment by CloudeLecaw** [comment]
i dont know if this really is a feature to support one of the 1000tns forks outside there.....

[13:44:52] [Server thread/INFO] []: Your server version is not compatible with this plugins version! Plugin will be disabled: v1_12_R1

This line tells you you cant use a plugin with this fork.

Since Magma is no official supported server of CMI (only Spigot and Paper are - no forks of them)
the change is low that you will get support from the dev there. he does not develop for the million other forks that are outside..

All you can hope is that the magma dev will make his server fork compatible to CMI.
or tell you the same as u saw in the line allready "Your server version is not compatible with this plugins version! Plugin will be disabled: v1_12_R1"

**Comment by IronworksProductions** [comment]
Is this just a matter of CMI disabling itself based on not recognizing the server core? because as i stated in the other issue it works almost perfectly on Mohist which is similar in design to magma. If that's the case maybe a flag to be able to override that autodisable at the understood risk that it may not work as designed would be useful.

**Comment by CloudeLecaw** [comment]
But its a fork, what means it doent need to be supported. I dont kniw much plugins that allows to disable this kind of check...
Why cmi should has this? If i wanne use a tool made for Windows nt-design and disabling in win 9x then. You can try to run it but have to except if it disables on start... Well hat tried to run CMI, niw its his part to except that magma and other forks are not spigot or PaperMC, they are forks

**Comment by RupticMC** [comment]
Seems to boot up just fine on Magma version 2c7ce42.

**Comment by CloudeLecaw** [comment]
https://github.com/magmafoundation/Magma

The description tells us. Cmi will not be erite fixed for this and also not Support this. If its working with these forks.. Fine... But dont aspect fixes caused by using this forks. And like allready told if your using these forks then your willing and acceptting that it might not work and that there will beno fixed for this.

This ticket can be closed now 😅

**Comment by IronworksProductions** [comment]
I can actually counter that logic. Windows NT and Windows 9x are both the same Windows OS architecture 9x is just an older version. As for Magma and spigot, if it follows spigot or paper spigot which it is for the version of MC then its not a matter of it being a fork, its a matter of this plugin not understanding or misunderstanding the environment that is created when forge is introduced. That being said, because that is the case, and your saying to have the magma devs fix it, would you be opposed to the magma devs patching cmi specifically to work with it like they did CoreProtect?

**Comment by CloudeLecaw** [comment]
the problem is the forks have changed code and thats the problem most ppl dont wanne see because everything seems to work with these forks but in 999999% there are codechanges the guys made to have so called "optimations" or give mod support or add new features, or or or
the forks will never be 100% equal to Spigot or Papermc and thats one of the main reason for msot of this issues
The Server have to modify their code to accpet the plugin if something does not work because on spigot/paper it works
Its not CMIs job to be support with forks

**Comment by IronworksProductions** [comment]
Yes I am aware of both of that, but the easiest way to fix a plugin issue that is only breaking because it doesn't recognize mods is to patch the plugin itself, which based on what I have seen the devs of magma have been more than willing to patch a plugin to make it work better on their core/fork. They already patched CoreProtect to work with their core so patching CMI wouldn't be an outlandish request given what it does and how much of a following the plugin has.

**Comment by CloudeLecaw** [comment]
> I have seen the devs of magma have been more than willing to patch a plugin to make it work better on their core/fork. 

good for them. 
maybe you can tell them too to fix their server for CMI too^^
lets look at the dependencies... hmm

![Snag_132c60e](https://user-images.githubusercontent.com/12080017/88866682-b3b9e880-d20b-11ea-96f5-bd40b65a4769.png)

Lett see where Magma is listet there... hmmmm?^^

You see -we can talk about this million times now.. but no support for magma and other forks if they have issues that papermc orr spigot does not have.

So finnal lets see what Zrips thinks about forks:
> Well if you have some completely different platform then it was said to be supported, then ether i need to add direct support for that build, which doesn't seam to happen any time soon as i cant keep track of all 50+ different forks anyone decides to create. Or developer of that fork could simply keep spigot/paper spigot methods without actually modifying or removing them...
> At this point i can only issue some crude patch to fail silently, where we will print out one error message and continue failing on that method silently after that.

you can fine this stament also here where a guy has console spam because of other fork
https://github.com/Zrips/CMI/issues/3829

Lets see what was the solution there...ooff they made codechanges in the fork... thats the right way ^^


nnow pls close this ticket :-)

**Comment by IronworksProductions** [comment]
Its not mine, otherwise I would have before we got into this discussion. I was just trying to ask questions and have a discussion not cause an argument.

**Comment by variananora** [comment]
Zrips probably wont do it, because it make too much hassle for him. (I guess.)

**Comment by CloudeLecaw** [comment]
> Zrips probably wont do it, because it make too much hassle for him. (I guess.)

I agree with this

**Comment by IronworksProductions** [comment]
> Well if you have some completely different platform then it was said to be supported, then ether I need to add direct support for that build, which doesn't seem to happen any time soon as I can't keep track of all 50+ different forks anyone decides to create. Or developer of that fork could simply keep spigot/paper spigot methods without actually modifying or removing them...
At this point, I can only issue some crude patch to fail silently, where we will print out one error message and continue failing on that method silently after that.

> 

While I understand the logic of this statement, forgive me for saying this but its heavily flawed when being used in this context. While there maybe 100s of pure spigot forks, there are only 2 bukkit forge implementations (3 if you count catserver) and both of them heavily derive from the same source material.(the last remnants of cauldron/thermos/kettle). Plus while it would be hassle in the start, the pay off of supporting those two branches would be immense. CMI is a paid plugin so it wouldn't exist or still be developed if people stopped paying for it. and there is a whole market of servers using either magma or mohist that would be very willing to use CMI if the dev wasn't so against making it work on those platforms.(which doesn't exactly get people to buy the plugin that use those server cores). Both Mohist and Magma have very documented API on how to get it to work and since they both derive from the same source, if you get one to work on magma, it is very _very_ likely it will also work on mohist as well. Plus if he does get it to work, that's more servers that can use it, which means more people who would then pay for the plugin, thus funding development. And before someone says I forgot about sponge, sponge doesn't implement bukkti at all, its API is entirely different. and while it does work well for mods and plugins, its different API has made the plugin market, _lacking_ to say the very least. spigot works great for vanilla servers but some server owners feel it doesn't provide enough power so they go to forge, but that doesn't have enough capability server side for running the server. bukkit forge implementations like magma and mohist offer the best of both worlds, allowing the power of both forge and spigot/paper to be combined to make a truly amazing and unique server. Which is why i think it should be at least considered instead of rejected just because its slightly different than spigot and paper vanilla.

**Comment by CloudeLecaw** [comment]
You should go to some brooker site and post this little advertise there. Sounds like a good advert for this fork ;) 
But i think Zrips allready know about this fakt. But each fork is Promotioned to be the best ;) I can understood the dev that he dont wanne Support the forks :) dont forget its still just a hobby in their free time. And i Dont think the main reason for creating this is making Profit. But who pais the hours of work he would need to get into it? Its not that easy like we user thinks all time. What would this mean for cmi? Well maybe Zrips need to offer 35€ for cmi because it cost tons of more time to be allways compatible with forks... Well at the moment it works for spigot and papermc. If your using other Serverfork on this price condishtions then accept issues and may breaks of features or maybe the whole plugin. Thats what is writen. You knewed before your purchased it that your wanne use it on other forks. So accept the results. By the way.. Not even the 3 bukkitt forks u mentioned are Supported. Just one fork is official Supported and this is papermc. Thats what the dev has decided. Accept his decission <3

**Comment by Tim-Rambo** [comment]
> You should go to some brooker site and post this little advertise there. Sounds like a good advert for this fork ;)
> But i think Zrips allready know about this fakt. But each fork is Promotioned to be the best ;) I can understood the dev that he dont wanne Support the forks :) dont forget its still just a hobby in their free time. And i Dont think the main reason for creating this is making Profit. But who pais the hours of work he would need to get into it? Its not that easy like we user thinks all time. What would this mean for cmi? Well maybe Zrips need to offer 35€ for cmi because it cost tons of more time to be allways compatible with forks... Well at the moment it works for spigot and papermc. If your using other Serverfork on this price condishtions then accept issues and may breaks of features or maybe the whole plugin. Thats what is writen. You knewed before your purchased it that your wanne use it on other forks. So accept the results. By the way.. Not even the 3 bukkitt forks u mentioned are Supported. Just one fork is official Supported and this is papermc. Thats what the dev has decided. Accept his decission <3


I bought CMI long time ago and knew that it would work on spigot/papermc

But when i made my modpack and was rly rly excited to finally use CMI for the first time bcs it provides so much so i dont need 10 plugins to do the same. i was excited for my modpack server but now i have to trash that and get other plugins, which i wanted to avoid.

And if CMI had an paid ADDON for making this work I WOULD 100% buy that addon for 20-30euro.

I got no problem with paying for an extra addon for cmi to make it work.

I Didnt know that cmi wouldnt work on magma... I was excited bcs magma is forge + plugins so i assumed it worked.

But when i tried it. It didnt work at all and i am upset and sad bcs of this.

I guess i just have to deal with it and move on...


My suggestion is to make an addon or to actually support it. more people will see the potetial in CMI there are so many features i want for my modpack server but i cant use those. and some of the features might not be out there at all. And i don`t want 10-20 plugins for my modpack server. than i can almost just scratch my modpack server and use forge mods to make what i want, which again is rly sad. I am upset about CMI not working with magma.

CMI would get more attetion into the modding community server owners would see the same as i see into CMI
I hope zrips gets patches to allow magma to work so CMI can reach more people out there not just from papermc and spigot community but also the modding community that has just recently started to get forks that work well with papermc/spigot + forge forks.

I hope it gets the support it needs and i am sure the magma dev would 100% work with zrips to help get the patches etc needed 👍 

Thanks
TimRambo

PS: Sorry about my spelling etc just too lazy to type well lol.
english isnt my first language :P

**Comment by Tim-Rambo** [comment]
Just to add to my comment

There are not a ton of forks with paper/spigot plugins + forge out there.

I recently discovered mohist and than a bit later magma.


The plugins + forge was all abounded in 1.7.10 with thermos and what not.


I dont take sponge into the picture bcs its not what i wanna use for my server at all.
I see more potetial in the forks that support forge + paper/spigot plugins

Thanks

**Comment by Tim-Rambo** [comment]
https://github.com/magmafoundation/Magma/issues/204

**Comment by Zrips** [comment]
To add 2cents to this conversation. I did some testing with this, well, build from https://magmafoundation.org seams to be working while https://github.com/magmafoundation/Magma/releases/tag/stable-1 seams to be broken.
This is with that issue where CMI disables itself. Tested on this broken version where it CMI failed to load itself, issue is that it doesn't indicate what went sideways and why it created error, it simply spits out NPE one liner which says nothing informational. Tried to get actual error in multiple ways, still nothing, it simply failed and who knows why.
But to be on same page, which exact version you used? Like a download link.

**Comment by Tim-Rambo** [comment]
> To add 2cents to this conversation. I did some testing with this, well, build from https://magmafoundation.org seams to be working while https://github.com/magmafoundation/Magma/releases/tag/stable-1 seams to be broken.
> This is with that issue where CMI disables itself. Tested on this broken version where it CMI failed to load itself, issue is that it doesn't indicate what went sideways and why it created error, it simply spits out NPE one liner which says nothing informational. Tried to get actual error in multiple ways, still nothing, it simply failed and who knows why.
> But to be on same page, which exact version you used? Like a download link.

Magma Version: This server is running Magma version 2798a77 (Implementing API version 1.12.2-R0.1, Forge Version 14.23.5.2838) thats the version that i used.

**Comment by IronworksProductions** [comment]
@Zrips https://github.com/magmafoundation/Magma/releases/tag/v2c7ce42
thats the latest release for magma. You actually used one of the really really old builds when trying to test

**Comment by RupticMC** [comment]
Almost as if I said that version worked two days ago. 😅

**Comment by IronworksProductions** [comment]
I had tested it myself with my server, and at least with my set up /cmi gives no response from console or as a player. and the commands connected to it fail silently as well. Though I have yet to figure out the root cause of that bug

**Comment by IronworksProductions** [comment]
I also am getting NPE's from plan on both mohist and magma within the first few seconds of the server running which disables the plans ability to track tps

**Comment by RupticMC** [comment]
Plan seems to run normally for me. Do you use NetworkManager on the server by chance? I cannot say that NM breaks CMI and Plan, but I noticed it broke some of my other plugins that I was using before I started using Plan and CMI.

**Comment by IronworksProductions** [comment]
No, this is the exact list of plugins I am using: §aAdvancedAchievements§f, §aBoss§f, §aBuycraftX§f, §aChatControl§f, §aChestShop§f, §aChestShop-LegacyIds§f, §aCitizens§f, §aCMI§f, §aCoreProtect§f, §aCratesPlus§f, §aCustomItems§f, §aDenizen§f, §aDiscordSRV§f, §aFactions§f, §aHolograms§f, §aInventoryRollback§f, §aIslandCraft§f, §aKnokkoCore§f, §aLuckPerms§f, §aMassiveCore§f, §aMultiWorld§f, §aPlan§f, §aReact§f, §aSentinel§f, §aServerListPlus§f, §aSuperbVote§f, §aTheNewEconomy§f, §aUltimateAutoRestart§f, §aUltimateTroll§f, §aVault§f, §aVotifier§f, §aWorldEdit

**Comment by IronworksProductions** [comment]
Any of those pop out to you as ones that would bug out? Boss I know crashes hilariously on magma but I doubt that's related

**Comment by RupticMC** [comment]
The only thing that points out to me is any of the "Massive" plugins because I believe it is completely abandoned. Also, if Boss is having a problem, might as well remove it. There is some plugins on there that I absolutely do not know of. 

I Suggest removing all plugins but leave in CMU, Buycraft, DiscordSRV, ChatControl, LuckPerms, Votifier, WorldEdit, and Vault. See if it loads up normally with those. If it does, its one of the plugins that I did NOT list and I would continue to put a couple plugins at a time onto the server until it stops working. 

If it does not work with all the plugins I listed above, remove them until you get to ONLY CMI on the server. If it still does not work, then it could be a mod for all I know.

**Comment by wolflord212312** [comment]
No magma can not fix it there self since this is a clossed source plugin nor can cauldron inless this becomes open source there will be no fix so the plugin dev either give the magma dev the code along with the cauldron dev or this has to be fixed plugin side.
